File: C_TopInterviewQuestions-HardCollection/a_array_and_strings/i_basic_calculator_ii.rs
struct Solution {}

impl Solution {
    // ^\d+( *[\+\-*/] *\d+)$
    pub fn calculate(s: String) -> i32 {
        let mut result = 0i32;
        let mut memory = 0i32;
        let mut number = 0i32;

        // save the first number to memory
        let mut operator = '+';

        let calc = |result: &mut i32, memory: &mut i32, number: &i32, operator: &char| {
            match operator {
                '+' => {
                    *result += *memory;
                    *memory = *number;
                }
                '-' => {
                    *result = *result + *memory;
                    *memory = -*number;
                }
                '*' => {
                    *memory = *memory * *number;
                }
                '/' => {
                    *memory = *memory / *number;
                }
                _ => {
                    panic!()
                }
            }
        };

        for c in s.chars() {
            match c {
                '0'..='9' => {
                    number = number * 10 + c as i32 - 48;
                }
                ' ' => {}
                _ => {
                    calc(&mut result, &mut memory, &number, &operator);
                    number = 0;
                    operator = c;
                }
            }
        }

        // calc the last operator
        calc(&mut result, &mut memory, &number, &operator);
        // maybe there is still some memory, remember it
        return result + memory;
    }
}

fn main() {
    println!("{}", 7 == Solution::calculate(String::from("3+2*2")));
    println!("{}", 1 == Solution::calculate(String::from(" 3/2 ")));
    println!("{}", 5 == Solution::calculate(String::from(" 3+5 / 2 ")));
    println!("{}", 18 == Solution::calculate(String::from(" 18")));
    println!("{}", 18 == Solution::calculate(String::from(" 33-15")));
}

File: C_TopInterviewQuestions-HardCollection/a_array_and_strings/i_basic_calculator_ii.ts
// ^\d+( *[\+\-*/] *\d+)$
function calculate(s: string): number {
    let result = 0
    let memory = 0
    let number = 0

    // at the very beginning, just save the first number to memory
    let operator = "+"
    const calc = () => {
        if (operator == "+") {
            result += memory
            memory = number
        } else if (operator == "-") {
            result = result + memory
            memory = -number
        } else if (operator == "*") {
            memory = memory * number
        } else if (operator == "/") {
            // bigint has performance issue
            // so calc with number and ignore the fractional part
            memory = ~~(memory / number)
        } else {
            throw new Error("unexpected rune")
        }
    }

    for (const c of s) {
        if (c >= "0" && c <= "9") {
            number = number * 10 + Number(c)
        } else if (c != " ") {
            calc()
            number = 0
            operator = c
        }
    }

    // since there is no operator after the last number
    // calc the last time manually
    calc()

    // maybe there is still some memory, remember it
    return result + memory
}

console.log(7 == calculate("3+2*2"))
console.log(1 == calculate(" 3/2 "))
console.log(5 == calculate(" 3+5 / 2 "))
console.log(18 == calculate(" 18"))
console.log(18 == calculate(" 33-15"))
console.log(13 == calculate("14-3/2"))
